// grunt-ts/tasks/ts.js
/// <reference path="../defs/tsd.d.ts"/>
/// <reference path="./modules/interfaces.d.ts"/>
/// <reference path="../defs/csproj2ts/csproj2ts.d.ts" />
var _ = require('lodash');
var path = require('path');
var fs = require('fs');
var csproj2ts = require('csproj2ts');
var utils = require('./modules/utils');
var compileModule = require('./modules/compile');
var referenceModule = require('./modules/reference');
var amdLoaderModule = require('./modules/amdLoader');
var html2tsModule = require('./modules/html2ts');
var templateCacheModule = require('./modules/templateCache');
var transformers = require('./modules/transformers');
var Promise = require('es6-promise').Promise;
function timeIt(makeIt) {
    var starttime = new Date().getTime();
    var it = makeIt();
    var endtime = new Date().getTime();
    return {
        it: it,
        time: endtime - starttime
    };
}
function asyncSeries(arr, iter) {
    arr = arr.slice(0);
    var memo = [];
    return new Promise(function (resolve, reject) {
        var next = function () {
            if (arr.length === 0) {
                resolve(memo);
                return;
            }
            Promise.cast(iter(arr.shift())).then(function (res) {
                memo.push(res);
                next();
            }, reject);
        };
        next();
    });
}
function pluginFn(grunt) {
    /////////////////////////////////////////////////////////////////////
    // The grunt task
    ////////////////////////////////////////////////////////////////////
    grunt.registerMultiTask('ts', 'Compile TypeScript files', function () {
        var currenttask = this;
        var done = currenttask.async();
        var watch;
        var filesCompilationIndex = 0;
        var options = currenttask.options({
            allowBool: false,
            allowImportModule: false,
            compile: true,
            declaration: false,
            emitDecoratorMetadata: false,
            mapRoot: '',
            module: null,
            noImplicitAny: false,
            noResolve: false,
            comments: null,
            removeComments: null,
            sourceMap: true,
            sourceRoot: '',
            target: 'es5',
            verbose: false,
            fast: 'watch',
            compiler: '',
            htmlOutputTemplate: null,
            htmlModuleTemplate: '<%= filename %>',
            htmlVarTemplate: '<%= ext %>',
            htmlOutDir: null,
            htmlOutDirFlatten: false,
            failOnTypeErrors: true,
            noEmitOnError: false,
            preserveConstEnums: false,
            suppressImplicitAnyIndexErrors: false,
            noEmit: false,
            inlineSources: false,
            inlineSourceMap: false,
            newLine: utils.eol,
            isolatedModules: false,
            noEmitHelpers: false,
            additionalFlags: ''
        });
        var rawTaskOptions = (grunt.config.getRaw(currenttask.name + '.options') || {});
        var rawTargetConfig = (grunt.config.getRaw(currenttask.name + '.' + currenttask.target) || {});
        var rawTargetOptions = (grunt.config.getRaw(currenttask.name + '.' + currenttask.target + '.options') || {});
        var vs = getVSSettings(rawTargetConfig);
        if (vs) {
            csproj2ts.getTypeScriptSettings({
                ProjectFileName: vs.project,
                ActiveConfiguration: vs.config || undefined
            }).then(function (vsConfig) {
                proceed(vsConfig);
            }).catch(function (error) {
                var errormessage;
                if (error.errno === 34) {
                    errormessage = 'In task "' + currenttask.target + '" - could not find VS project at "' + error.path + '".';
                }
                else {
                    errormessage = 'In task "' + currenttask.target + '".  Error #' + error.errno + '.  ' + error;
                }
                grunt.fail.warn(errormessage, error.errno);
                done(error);
            });
        }
        else {
            proceed();
        }
        function proceed(vsProjectTypeScriptSettings) {
            if (vsProjectTypeScriptSettings && !vs.ignoreSettings) {
                options.declaration = utils.firstElementWithValue([vsProjectTypeScriptSettings.GeneratesDeclarations,
                    options.declaration]);
                options.mapRoot = utils.firstElementWithValue([vsProjectTypeScriptSettings.MapRoot,
                    options.mapRoot]);
                options.module = utils.firstElementWithValue([vsProjectTypeScriptSettings.ModuleKind,
                    options.module]);
                if (options.module === 'none') {
                    options.module = null;
                }
                options.noEmitOnError = utils.firstElementWithValue([vsProjectTypeScriptSettings.NoEmitOnError,
                    options.noEmitOnError]);
                options.noImplicitAny = utils.firstElementWithValue([vsProjectTypeScriptSettings.NoImplicitAny,
                    options.noImplicitAny]);
                options.noResolve = utils.firstElementWithValue([vsProjectTypeScriptSettings.NoResolve,
                    options.noResolve]);
                rawTargetConfig.outDir = utils.firstElementWithValue([vsProjectTypeScriptSettings.OutDir,
                    rawTargetConfig.outDir]);
                rawTargetConfig.out = utils.firstElementWithValue([vsProjectTypeScriptSettings.OutFile,
                    rawTargetConfig.out]);
                options.preserveConstEnums = utils.firstElementWithValue([vsProjectTypeScriptSettings.PreserveConstEnums,
                    options.preserveConstEnums]);
                options.removeComments = utils.firstElementWithValue([vsProjectTypeScriptSettings.RemoveComments,
                    options.removeComments]);
                if (options.removeComments) {
                    options.comments = null;
                }
                else {
                    options.comments = true;
                }
                options.sourceMap = utils.firstElementWithValue([vsProjectTypeScriptSettings.SourceMap,
                    options.sourceMap]);
                options.sourceRoot = utils.firstElementWithValue([vsProjectTypeScriptSettings.SourceRoot,
                    options.sourceRoot]);
                options.suppressImplicitAnyIndexErrors = utils.firstElementWithValue([vsProjectTypeScriptSettings.SuppressImplicitAnyIndexErrors,
                    options.suppressImplicitAnyIndexErrors]);
                options.target = utils.firstElementWithValue([vsProjectTypeScriptSettings.Target,
                    options.target]);
            }
            var srcFromVS_RelativePathsFromGruntFile = [];
            if (vsProjectTypeScriptSettings) {
                var absolutePathToVSProjectFolder = path.resolve(vsProjectTypeScriptSettings.VSProjectDetails.ProjectFileName, '..');
                if (!vs.ignoreFiles) {
                    _.map(_.uniq(vsProjectTypeScriptSettings.files), function (file) {
                        var absolutePathToFile = path.normalize(path.join(absolutePathToVSProjectFolder, file));
                        var relativePathToFile = path.relative(path.resolve('.'), absolutePathToFile).replace(new RegExp('\\' + path.sep, 'g'), '/');
                        if (srcFromVS_RelativePathsFromGruntFile.indexOf(relativePathToFile) === -1 &&
                            currenttask.filesSrc.indexOf(relativePathToFile) === -1) {
                            srcFromVS_RelativePathsFromGruntFile.push(relativePathToFile);
                        }
                    });
                    if (srcFromVS_RelativePathsFromGruntFile.length > 0) {
                        currenttask.files.push({ src: srcFromVS_RelativePathsFromGruntFile });
                    }
                }
                if (!vs.ignoreSettings) {
                    if (vsProjectTypeScriptSettings.OutDir) {
                        rawTargetConfig.outDir = path.relative(path.resolve('.'), path.normalize(path.join(absolutePathToVSProjectFolder, vsProjectTypeScriptSettings.OutDir)));
                    }
                    if (vsProjectTypeScriptSettings.OutFile) {
                        rawTargetConfig.out = path.relative(path.resolve('.'), path.normalize(path.join(absolutePathToVSProjectFolder, vsProjectTypeScriptSettings.OutFile)));
                    }
                }
            }
            options.htmlOutputTemplate = rawTargetOptions.htmlOutputTemplate || rawTaskOptions.htmlOutputTemplate;
            var lineEndingToUse = utils.newLineParameterAsActual(rawTargetOptions.newLine) ||
                utils.newLineParameterAsActual(rawTaskOptions.newLine) ||
                utils.eol;
            options.newLine = rawTargetOptions.newLine || rawTaskOptions.newLine ||
                utils.newLineActualAsParameter(utils.eol);
            options.htmlModuleTemplate = rawTargetOptions.htmlModuleTemplate || rawTaskOptions.htmlModuleTemplate;
            options.htmlVarTemplate = rawTargetOptions.htmlVarTemplate || rawTaskOptions.htmlVarTemplate;
            options.htmlOutDir = rawTargetConfig.htmlOutDir;
            options.htmlOutDirFlatten = rawTargetConfig.htmlOutDirFlatten;
            options.isolatedModules = rawTargetOptions.isolatedModules || rawTaskOptions.isolatedModules;
            options.noEmitHelpers = rawTargetOptions.noEmitHelpers || rawTaskOptions.noEmitHelpers;
            options.additionalFlags = utils.firstElementWithValue([rawTargetOptions.additionalFlags, rawTaskOptions.additionalFlags]);
            options.allowBool = 'allowbool' in options ?
                options['allowbool'] : options.allowBool;
            options.allowImportModule = 'allowimportmodule' in options ?
                options['allowimportmodule'] : options.allowImportModule;
            options.sourceMap = 'sourcemap' in options ?
                options['sourcemap'] : options.sourceMap;
            options.emitDecoratorMetadata = 'emitdecoratormetadata' in options ?
                options['emitdecoratormetadata'] : options.emitDecoratorMetadata;
            options.noEmit = 'noemit' in options ?
                options['noemit'] : options.noEmit;
            options.inlineSources = 'inlinesources' in options ?
                options['inlinesources'] : options.inlineSources;
            options.inlineSourceMap = 'inlinesourcemap' in options ?
                options['inlinesourcemap'] : options.inlineSourceMap;
            options.isolatedModules = 'isolatedmodules' in options ?
                options['isolatedmodules'] : options.isolatedModules;
            options.noEmitHelpers = 'noemithelpers' in options ?
                options['noemithelpers'] : options.noEmitHelpers;
            options.additionalFlags = 'additionalflags' in options ?
                options['additionalflags'] : options.additionalFlags;
            if (options.fast !== 'watch' && options.fast !== 'always' && options.fast !== 'never') {
                console.warn(('"fast" needs to be one of : "watch" (default) | "always" | "never" but you provided: ' + options.fast).magenta);
                if (currenttask.files) {
                    options.fast = 'never';
                }
                else {
                    options.fast = 'watch';
                }
            }
            if ((options.fast === 'watch' || options.fast === 'always') && rawTargetConfig.files) {
                grunt.log.writeln(('Warning: Task "' + currenttask.target +
                    '" is attempting to use fast compilation with "files".  This is not currently supported.  Setting "fast" to "never".').magenta);
                options.fast = 'never';
            }
            logBadConfigWithFiles(rawTargetConfig, currenttask, rawTargetOptions);
            if (!options.htmlModuleTemplate) {
                options.htmlModuleTemplate = '<%= filename %>';
            }
            if (!options.htmlVarTemplate) {
                options.htmlVarTemplate = '<%= ext %>';
            }
            if (!options.htmlOutDir) {
                options.htmlOutDir = null;
            }
            if (!options.htmlOutDirFlatten) {
                options.htmlOutDirFlatten = false;
            }
            if (options.removeComments === null) {
                options.removeComments = !options.comments;
            }
            else if (options.comments !== null && !vs) {
                console.warn('WARNING: Option "comments" and "removeComments" should not be used together'.magenta);
                if (options.removeComments === options.comments) {
                    console.warn('Either option will suffice (and removing the other will have no effect).'.magenta);
                }
                else {
                    console.warn(('The --removeComments value of "' + options.removeComments + '" ' +
                        'supercedes the --comments value of "' + options.comments + '"').magenta);
                }
            }
            options.removeComments = !!options.removeComments;
            if (options.inlineSources) {
                options.inlineSourceMap = true;
            }
            if (options.inlineSourceMap) {
                options.sourceMap = true;
            }
            if (currenttask.files.length === 0 && rawTargetOptions.compile) {
                grunt.log.writeln('Zero files found to compile in target "' + currenttask.target + '". Compilation will be skipped.');
            }
            asyncSeries(currenttask.files, function (target) {
                var reference = processIndividualTemplate(rawTargetConfig.reference);
                var referenceFile;
                var referencePath;
                if (!!reference) {
                    referenceFile = path.resolve(reference);
                    referencePath = path.dirname(referenceFile);
                }
                function isReferenceFile(filename) {
                    return path.resolve(filename) === referenceFile;
                }
                function fetchTargetOutOrElseTryTargetDest(target) {
                    var targetout = target.out;
                    if (!targetout) {
                        if (target.dest) {
                            targetout = utils.getOrGetFirst(target.dest);
                        }
                        else if (target.files) {
                            var filesKeys = _.keys(target.files);
                            if (filesKeys.length > filesCompilationIndex &&
                                utils.isJavaScriptFile(filesKeys[filesCompilationIndex])) {
                                targetout = filesKeys[filesCompilationIndex];
                            }
                            else if (filesKeys.length > filesCompilationIndex &&
                                target.files[filesKeys[filesCompilationIndex]].dest) {
                                targetout = utils.getOrGetFirst(target.files[filesKeys[filesCompilationIndex]].dest);
                            }
                        }
                    }
                    return targetout;
                }
                var outFile = fetchTargetOutOrElseTryTargetDest(rawTargetConfig);
                var outFile_d_ts;
                if (!!outFile) {
                    outFile = path.resolve(outFile);
                    outFile_d_ts = outFile.replace('.js', '.d.ts');
                }
                function isOutFile(filename) {
                    return path.resolve(filename) === outFile_d_ts;
                }
                function isBaseDirFile(filename, targetFiles) {
                    var baseDirFile = '.baseDir.ts';
                    var bd = '';
                    if (!rawTargetConfig.baseDir) {
                        bd = utils.findCommonPath(targetFiles, '/');
                        rawTargetConfig.baseDir = bd;
                    }
                    return path.resolve(filename) === path.resolve(path.join(bd, baseDirFile));
                }
                var amdloader = rawTargetConfig.amdloader;
                var amdloaderFile;
                var amdloaderPath;
                if (!!amdloader) {
                    amdloaderFile = path.resolve(amdloader);
                    amdloaderPath = path.dirname(amdloaderFile);
                }
                processAllTargetTemplates(rawTargetConfig, rawTargetOptions);
                outFile = processIndividualTemplate(outFile);
                function runCompilation(files, target, options) {
                    grunt.log.writeln('Compiling...'.yellow);
                    var filesToCompile = files;
                    var starttime = new Date().getTime();
                    var endtime;
                    return compileModule.compileAllFiles(filesToCompile, target, options, currenttask.target, outFile)
                        .then(function (result) {
                        endtime = new Date().getTime();
                        grunt.log.writeln('');
                        if (!result) {
                            grunt.log.error('Error: No result from tsc.'.red);
                            return false;
                        }
                        if (result.code === 8) {
                            grunt.log.error('Error: Node was unable to run tsc.  Possibly it could not be found?'.red);
                            return false;
                        }
                        var isError = (result.code !== 0);
                        var level1ErrorCount = 0, level5ErrorCount = 0, nonEmitPreventingWarningCount = 0;
                        var hasTS7017Error = false;
                        var hasPreventEmitErrors = _.foldl(result.output.split('\n'), function (memo, errorMsg) {
                            var isPreventEmitError = false;
                            if (errorMsg.search(/error TS7017:/g) >= 0) {
                                hasTS7017Error = true;
                            }
                            if (errorMsg.search(/error TS1\d+:/g) >= 0) {
                                level1ErrorCount += 1;
                                isPreventEmitError = true;
                            }
                            else if (errorMsg.search(/error TS5\d+:/) >= 0) {
                                level5ErrorCount += 1;
                                isPreventEmitError = true;
                            }
                            else if (errorMsg.search(/error TS\d+:/) >= 0) {
                                nonEmitPreventingWarningCount += 1;
                            }
                            return memo || isPreventEmitError;
                        }, false) || false;
                        var isOnlyTypeErrors = !hasPreventEmitErrors;
                        if (hasTS7017Error) {
                            grunt.log.writeln(('Note:  You may wish to enable the suppressImplicitAnyIndexErrors' +
                                ' grunt-ts option to allow dynamic property access by index.  This will' +
                                ' suppress TypeScript error TS7017.').magenta);
                        }
                        if (level1ErrorCount + level5ErrorCount + nonEmitPreventingWarningCount > 0) {
                            if ((level1ErrorCount + level5ErrorCount > 0) || options.failOnTypeErrors) {
                                grunt.log.write(('>> ').red);
                            }
                            else {
                                grunt.log.write(('>> ').green);
                            }
                            if (level5ErrorCount > 0) {
                                grunt.log.write(level5ErrorCount.toString() + ' compiler flag error' +
                                    (level5ErrorCount === 1 ? '' : 's') + '  ');
                            }
                            if (level1ErrorCount > 0) {
                                grunt.log.write(level1ErrorCount.toString() + ' syntax error' +
                                    (level1ErrorCount === 1 ? '' : 's') + '  ');
                            }
                            if (nonEmitPreventingWarningCount > 0) {
                                grunt.log.write(nonEmitPreventingWarningCount.toString() +
                                    ' non-emit-preventing type warning' +
                                    (nonEmitPreventingWarningCount === 1 ? '' : 's') + '  ');
                            }
                            grunt.log.writeln('');
                            if (isOnlyTypeErrors && !options.failOnTypeErrors) {
                                grunt.log.write(('>> ').green);
                                grunt.log.writeln('Type errors only.');
                            }
                        }
                        var isSuccessfulBuild = (!isError ||
                            (isError && isOnlyTypeErrors && !options.failOnTypeErrors));
                        if (isSuccessfulBuild) {
                            var time = (endtime - starttime) / 1000;
                            grunt.log.writeln('');
                            grunt.log.writeln(('TypeScript compilation complete: ' + time.toFixed(2) +
                                's for ' + result.fileCount + ' typescript files').green);
                        }
                        else {
                            grunt.log.error(('Error: tsc return code: ' + result.code).yellow);
                        }
                        return isSuccessfulBuild;
                    }).catch(function (err) {
                        grunt.log.writeln(('Error: ' + err).red);
                        return false;
                    });
                }
                function filterFilesAndCompile() {
                    var filesToCompile = [];
                    if (currenttask.data.src || currenttask.data.vs) {
                        if (currenttask.data.src) {
                            filesToCompile = grunt.file.expand(currenttask.data.src);
                        }
                        _.map(srcFromVS_RelativePathsFromGruntFile, function (file) {
                            if (filesToCompile.indexOf(file) === -1) {
                                filesToCompile.push(file);
                            }
                        });
                    }
                    else {
                        if (_.isArray(currenttask.data.files)) {
                            filesToCompile = grunt.file.expand(currenttask.data.files[filesCompilationIndex].src);
                        }
                        else if (currenttask.data.files[target.dest]) {
                            filesToCompile = grunt.file.expand(currenttask.data.files[target.dest]);
                        }
                        else {
                            filesToCompile = grunt.file.expand([currenttask.data.files.src]);
                        }
                        filesCompilationIndex += 1;
                    }
                    filesToCompile = filesToCompile.filter(function (file) {
                        var stats = fs.lstatSync(file);
                        return !stats.isDirectory() && !isOutFile(file) && !isBaseDirFile(file, filesToCompile);
                    });
                    var generatedFiles = [];
                    if (currenttask.data.html) {
                        var html2tsOptions = {
                            moduleFunction: _.template(options.htmlModuleTemplate),
                            varFunction: _.template(options.htmlVarTemplate),
                            htmlOutputTemplate: options.htmlOutputTemplate,
                            htmlOutDir: options.htmlOutDir,
                            flatten: options.htmlOutDirFlatten,
                            eol: lineEndingToUse
                        };
                        var htmlFiles = grunt.file.expand(currenttask.data.html);
                        generatedFiles = _.map(htmlFiles, function (filename) { return html2tsModule.compileHTML(filename, html2tsOptions); });
                    }
                    if (currenttask.data.templateCache) {
                        if (!currenttask.data.templateCache.src || !currenttask.data.templateCache.dest || !currenttask.data.templateCache.baseUrl) {
                            grunt.log.writeln('templateCache : src, dest, baseUrl must be specified if templateCache option is used'.red);
                        }
                        else {
                            var templateCacheSrc = grunt.file.expand(currenttask.data.templateCache.src);
                            var templateCacheDest = path.resolve(rawTargetConfig.templateCache.dest);
                            var templateCacheBasePath = path.resolve(rawTargetConfig.templateCache.baseUrl);
                            templateCacheModule.generateTemplateCache(templateCacheSrc, templateCacheDest, templateCacheBasePath, lineEndingToUse);
                        }
                    }
                    if (!!referencePath) {
                        var result = timeIt(function () {
                            return referenceModule.updateReferenceFile(filesToCompile.filter(function (f) { return !isReferenceFile(f); }), generatedFiles, referenceFile, referencePath, lineEndingToUse);
                        });
                        if (result.it === true) {
                            grunt.log.writeln(('Updated reference file (' + result.time + 'ms).').green);
                        }
                    }
                    if (!!amdloaderPath) {
                        var referenceOrder = amdLoaderModule.getReferencesInOrder(referenceFile, referencePath, generatedFiles);
                        amdLoaderModule.updateAmdLoader(referenceFile, referenceOrder, amdloaderFile, amdloaderPath, rawTargetConfig.outDir);
                    }
                    transformers.transformFiles(filesToCompile, filesToCompile, rawTargetConfig, options, lineEndingToUse);
                    if (options.compile) {
                        if (filesToCompile.length > 0) {
                            return runCompilation(filesToCompile, rawTargetConfig, options).then(function (success) {
                                return success;
                            });
                        }
                        else {
                            grunt.log.writeln('No files to compile'.red);
                            return Promise.resolve(true);
                        }
                    }
                    else {
                        return Promise.resolve(true);
                    }
                }
                var lastCompile = 0;
                if (rawTargetConfig.files && rawTargetConfig.watch) {
                    grunt.log.writeln(('WARNING: Use of "files" with "watch" in target ' + currenttask.target +
                        ' is not supported in grunt-ts.  The "watch" will be ignored.  Use "src", or use grunt-contrib-watch' +
                        ' if you really do need to use "files".').magenta);
                }
                watch = rawTargetConfig.watch;
                if (!!watch) {
                    function handleFileEvent(filepath, displaystr, addedOrChanged) {
                        if (addedOrChanged === void 0) { addedOrChanged = false; }
                        if (!utils.endsWith(filepath.toLowerCase(), '.ts') && !utils.endsWith(filepath.toLowerCase(), '.html')) {
                            return;
                        }
                        if ((new Date().getTime() - lastCompile) <= 100) {
                            return;
                        }
                        grunt.log.writeln((displaystr + ' >>' + filepath).yellow);
                        filterFilesAndCompile();
                    }
                    var watchpath = grunt.file.expand(watch);
                    var chokidar = require('chokidar');
                    var watcher = chokidar.watch(watchpath, { ignoreInitial: true, persistent: true });
                    grunt.log.writeln(('Watching all TypeScript / Html files under : ' + watchpath).cyan);
                    watcher
                        .on('add', function (path) {
                        handleFileEvent(path, '+++ added   ', true);
                        lastCompile = new Date().getTime();
                    })
                        .on('change', function (path) {
                        handleFileEvent(path, '### changed ', true);
                        lastCompile = new Date().getTime();
                    })
                        .on('unlink', function (path) {
                        handleFileEvent(path, '--- removed ');
                        lastCompile = new Date().getTime();
                    })
                        .on('error', function (error) {
                        console.error('Error happened in chokidar: ', error);
                    });
                }
                lastCompile = new Date().getTime();
                return filterFilesAndCompile();
            }).then(function (res) {
                if (!watch) {
                    if (res.some(function (succes) {
                        return !succes;
                    })) {
                        done(false);
                    }
                    else {
                        done();
                    }
                }
            }, done);
        }
    });
    function logBadConfigWithFiles(config, task, targetOpt) {
        if (config.files) {
            if (config.vs) {
                grunt.log.writeln(('Warning: In task "' + task.target +
                    '", either "files" or "vs" should be used - not both.').magenta);
                return;
            }
            if (config.src) {
                grunt.log.writeln(('Warning: In task "' + task.target +
                    '", either "files" or "src" should be used - not both.').magenta);
                return;
            }
            if (config.out) {
                grunt.log.writeln(('Warning: In task "' + task.target +
                    '", either "files" or "out" should be used - not both.').magenta);
                return;
            }
            if (config.outDir) {
                grunt.log.writeln(('Warning: In task "' + task.target +
                    '", either "files" or "outDir" should be used - not both.').magenta);
                return;
            }
        }
        else {
            if (!config.src && !config.vs && targetOpt.compile) {
                grunt.log.writeln(('Warning: In task "' + task.target +
                    '", neither "files" nor "src" nor "vs" is specified.  Nothing will be compiled.').magenta);
                return;
            }
        }
    }
    function processAllTargetTemplates(targetCfg, targetOpt) {
        targetCfg.out = processIndividualTemplate(targetCfg.out);
        targetCfg.outDir = processIndividualTemplate(targetCfg.outDir);
        targetCfg.reference = processIndividualTemplate(targetCfg.reference);
        targetOpt.mapRoot = processIndividualTemplate(targetOpt.mapRoot);
        targetOpt.sourceRoot = processIndividualTemplate(targetOpt.sourceRoot);
    }
    function processIndividualTemplate(template) {
        if (template) {
            return grunt.template.process(template, {});
        }
        return template;
    }
    function getVSSettings(rawTargetOptions) {
        var vs = null;
        if (rawTargetOptions.vs) {
            var targetvs = rawTargetOptions.vs;
            if (typeof targetvs === 'string') {
                vs = {
                    project: targetvs,
                    config: '',
                    ignoreFiles: false,
                    ignoreSettings: false
                };
            }
            else {
                vs = {
                    project: targetvs.project || '',
                    config: targetvs.config || '',
                    ignoreFiles: targetvs.ignoreFiles || false,
                    ignoreSettings: targetvs.ignoreSettings || false
                };
            }
        }
        return vs;
    }
}
module.exports = pluginFn;
